ndc-reference/bin/reference/main.rs

```diff
@@ -1488,6 +1510,81 @@ fn eval_comparison_value(
     }
 }
 // ANCHOR_END: eval_comparison_value
+// ANCHOR: eval_nested_field
+fn eval_nested_field(
+    collection_relationships: &BTreeMap<String, models::Relationship>,
+    variables: &BTreeMap<String, serde_json::Value>,
+    state: &AppState,
+    value: Value,
+    nested_field: &models::NestedField,
+) -> Result<models::RowFieldValue> {
+    match nested_field {
+        models::NestedField::Object(models::NestedObject { fields }) => {
+            let full_row: Row = serde_json::from_value(value).map_err(|_| {
+                (
+                    StatusCode::BAD_REQUEST,
+                    Json(models::ErrorResponse {
+                        message: "Expected object".into(),
+                        details: serde_json::Value::Null,
+                    }),
+                )
+            })?;
+            let row = eval_row(
+                fields,
+                collection_relationships,
+                variables,
+                state,
+                &full_row,
+            )?;
+            Ok(models::RowFieldValue(serde_json::to_value(row).map_err(
+                |_| {
+                    (
+                        StatusCode::INTERNAL_SERVER_ERROR,
+                        Json(models::ErrorResponse {
+                            message: "Cannot encode rowset".into(),
+                            details: serde_json::Value::Null,
+                        }),
+                    )
+                },
+            )?))
+        }
+        models::NestedField::Array(models::NestedArray { fields }) => {
+            let array: Vec<Value> = serde_json::from_value(value).map_err(|_| {
+                (
+                    StatusCode::BAD_REQUEST,
+                    Json(models::ErrorResponse {
+                        message: "Expected array".into(),
+                        details: serde_json::Value::Null,
+                    }),
+                )
+            })?;
+            let result_array = match fields.deref() {
+                None => array
+                    .into_iter()
+                    .map(models::RowFieldValue)
+                    .collect::<Vec<_>>(),
+                Some(field) => array
+                    .into_iter()
+                    .map(|value| {
+                        eval_nested_field(collection_relationships, variables, state, value, field)
+                    })
+                    .collect::<Result<Vec<_>>>()?,
+            };
+            Ok(models::RowFieldValue(
```

Comment:
So we have two layers of `RowFieldValue` here, is that deliberate?

Reply:
Never mind, it was the `None` case that threw me off. Can we just return `array` directly and skip the reallocation?